### PR TITLE
[#161995] [Shared dev] [WIP] Duration based pricing - (5) Duration discounts applied to user reservation

### PR DESCRIPTION
**Pull request by LeticiaErrandonea**
# Release Notes

A one- or two-sentence explanation of what the change is and why, phrased for a non-technical audience. This should be copy-pasted as your commit message when the PR is merged. (Required)

Please preface your release note with "TECH TASK: " if the change is a technical task such as a gem upgrade or changes to a configuration file.

# Screenshot

Optional.

# Additional Context

Optional. Feel free to add/modify additional headers as appropriate, e.g. "Refactorings", "Concerns".

### CODE REVIEW

File: app/services/price_policies/time_based_price_calculator.rb [hunk 1]
```diff
@@ -40,6 +46,49 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
...
     end
 
-  end
+    def duration_pricing_cost(duration_mins)
+      duration_in_hours = duration_mins.to_f / 60
+
+      # 3 cases for interval_data:
+      # interval_data = [nil, { min_duration: 5, rate: 8 }]
+      # interval_data = [{ min_duration: 0, rate: 10 }, { min_duration: 5, rate: 8 }]
+      # interval_data = [{ min_duration: 0, rate: 10 }, nil]
```
Comment (LeticiaErrandonea) [inline app/services/price_policies/time_based_price_calculator.rb:55]: Plan on adding a comment that better explains this if we can't find a solution that's more self explanatory

File: app/services/price_policies/time_based_price_calculator.rb [hunk 2]
```diff
@@ -16,8 +16,14 @@ def initialize(price_policy)
     def calculate(start_at, end_at)
       return if start_at > end_at
       duration_mins = TimeRange.new(start_at, end_at).duration_mins
-      discount_multiplier = calculate_discount(start_at, end_at)
-      cost_and_subsidy(duration_mins, discount_multiplier)
+
+      if product.is_a?(Instrument) && product.duration_pricing_mode? && product.duration_rates.present?
+        # Should return something for subsidy as well?
```
Comment (LeticiaErrandonea) [inline app/services/price_policies/time_based_price_calculator.rb:21]: My guess is we'll need this to also return the `subsidy` and the calculation will either be `duration_in_hours * usage_subsidy` or just `0` if it doesn't make sense for this case

File: spec/services/price_policies/time_based_price_calculator_spec.rb [hunk 3]
```diff
@@ -3,196 +3,323 @@
 require "rails_helper"
 
 RSpec.describe PricePolicies::TimeBasedPriceCalculator do
-
```
Comment (LeticiaErrandonea) [inline spec/services/price_policies/time_based_price_calculator_spec.rb:6]: Relevant changes start in line 200. The rest are only moving the previous tests to a context

File: app/services/price_policies/time_based_price_calculator.rb (hunk 2 above)
Comment (giladshanan) [reply app/services/price_policies/time_based_price_calculator.rb:21]: I think that the return value needs to include subsidy, but I'm not clear if it should be 0 or something else

File: app/services/price_policies/time_based_price_calculator.rb [hunk 4]
```diff
@@ -40,6 +46,49 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
...
+
+      { cost: result[:cost] }
+    end
+
+    def build_intervals
+      sorted_duration_rates = product.duration_rates.sort_by { |dr| dr.min_duration || 1_000 }
+      lower_intervals = sorted_duration_rates.dup
+      higher_intervals = sorted_duration_rates.dup
+
+      if sorted_duration_rates.first.min_duration > 0
```
Comment (giladshanan) [inline app/services/price_policies/time_based_price_calculator.rb:85]: I think we want to require the first one to be 0

File: app/services/price_policies/time_based_price_calculator.rb [hunk 5]
```diff
@@ -40,6 +46,49 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
...
+      # interval_data = [nil, { min_duration: 5, rate: 8 }]
+      # interval_data = [{ min_duration: 0, rate: 10 }, { min_duration: 5, rate: 8 }]
+      # interval_data = [{ min_duration: 0, rate: 10 }, nil]
+      result = build_intervals.reduce({ time_left: duration_in_hours, cost: 0 }) do |acc, interval_data|
+        time_left = acc[:time_left]
+
+        interval_data_start = interval_data.first
+        interval_data_end = interval_data.last
+
+        hourly_rate = interval_data_start ? interval_data_start.rate : usage_rate * 60
```
Comment (giladshanan) [inline app/services/price_policies/time_based_price_calculator.rb:62]: dont need to use usage_rate here

File: app/services/price_policies/time_based_price_calculator.rb [hunk 6]
```diff
@@ -40,6 +46,49 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
...
+      lower_intervals = sorted_duration_rates.dup
+      higher_intervals = sorted_duration_rates.dup
+
+      if sorted_duration_rates.first.min_duration > 0
+        lower_intervals.prepend nil
+      else
+        higher_intervals.shift
+      end
+
+      lower_intervals.zip(higher_intervals)
```
Comment (giladshanan) [inline app/services/price_policies/time_based_price_calculator.rb:91]: can this return an array of hashes like [{interval_start: 0, interval_end: 5, rate: 10}]

File: app/services/price_policies/time_based_price_calculator.rb [hunk 7]
```diff
@@ -40,6 +46,49 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
...
+      result = build_intervals.reduce({ time_left: duration_in_hours, cost: 0 }) do |acc, interval_data|
+        time_left = acc[:time_left]
+
+        interval_data_start = interval_data.first
+        interval_data_end = interval_data.last
+
+        hourly_rate = interval_data_start ? interval_data_start.rate : usage_rate * 60
+
+        interval_start = interval_data_start&.min_duration || 0
+        interval_end = interval_data_end&.min_duration || Float::INFINITY
```
Comment (giladshanan) [inline app/services/price_policies/time_based_price_calculator.rb:65]: move `Float::INFINITY` to build_intervals method

File: app/services/price_policies/time_based_price_calculator.rb [hunk 8]
```diff
@@ -40,6 +46,49 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
...
+
+        time_to_charge = [time_left, interval_length].min
+
+        acc[:time_left] -= time_to_charge
+        acc[:cost] += hourly_rate * time_to_charge
 
+        acc
+      end
+
+      { cost: result[:cost] }
```
Comment (giladshanan) [inline app/services/price_policies/time_based_price_calculator.rb:77]: I think this should include subsidy: 0 to start

File: spec/factories/duration_rates.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :duration_rate do
+    min_duration { 3 }
```
Comment (giladshanan) [inline spec/factories/duration_rates.rb:5]: this should be 0

File: spec/services/price_policies/time_based_price_calculator_spec.rb [hunk 10]
```diff
@@ -3,196 +3,323 @@
...
 
+            it "uses duration rates rate" do
+              create(:duration_rate, instrument: product, min_duration: 0, rate: 60)
+
+              is_expected.to eq(cost: 60 + 15)
+            end
+          end
+        end
+
+        context "none of them with minimum duration of 0 hours" do
```
Comment (giladshanan) [inline spec/services/price_policies/time_based_price_calculator_spec.rb:243]: this case is going away

File: spec/services/price_policies/time_based_price_calculator_spec.rb [hunk 11]
```diff
@@ -3,196 +3,323 @@
...
+            let(:start_at) { Time.zone.local(2017, 4, 27, 12, 0) }
+            let(:end_at) { start_at + 1.hour + 15.minute }
 
+            it "uses duration rates rate" do
+              create(:duration_rate, instrument: product, min_duration: 0, rate: 60)
+
+              is_expected.to eq(cost: 60 + 15)
+            end
+          end
+        end
```
Comment (giladshanan) [inline spec/services/price_policies/time_based_price_calculator_spec.rb:241]: add a case for "Reservation time must be at least 1 minute within the time range in order to get the rate."

**Review by giladshanan — COMMENTED** [review]
Looks good!

**Pull request closed without merging**